File: Python/28-数组中出现次数超过一半的数字.py
# -*- coding:utf-8 -*-
class Solution:
    def MoreThanHalfNum_Solution(self, numbers):
        # write code here
        n = len(numbers)
        cur = numbers[0]
        cnt = 1
        for i in range(1, n):
            if numbers[i] == cur:
                cnt += 1
            else:
                cnt -= 1
                if cnt == 0:
                    cur = numbers[i]
                    cnt = 1
        if numbers.count(cur) > n // 2:
            return cur
        return 0

File: C++/28-数组中出现次数超过一半的数字.cpp
class Solution {
public:
    int MoreThanHalfNum_Solution(vector<int> numbers) {
        int cnt = 1, cur = numbers[0];
        for(int i = 1; i < numbers.size(); ++i){
            if(cur == numbers[i])
                cnt++;
            else{
                cnt--;
                if(cnt == 0){
                    cnt = 1;
                    cur = numbers[i];
                }
            }
        }
        cnt = 0;
        for(int i = 0; i < numbers.size(); ++i)
            if(numbers[i] == cur)
                cnt++;
        if(cnt > numbers.size() / 2)
            return cur;
        return 0;
    }
};